Zend/zend_compile.c

```diff
@@ -3018,17 +3018,29 @@ static zend_op *zend_delayed_compile_prop(znode *result, zend_ast *ast, uint32_t
 		 * check for a nullsafe access. */
 	} else {
 		zend_short_circuiting_mark_inner(obj_ast);
-		opline = zend_delayed_compile_var(&obj_node, obj_ast, type, 0);
+		/* Allow fetching constants by R, even in W context. */
+		uint32_t obj_type = obj_ast->kind == ZEND_AST_CONST ? BP_VAR_R : type;
+		opline = zend_delayed_compile_var(&obj_node, obj_ast, obj_type, 0);
 		if (opline && (opline->opcode == ZEND_FETCH_DIM_W
 				|| opline->opcode == ZEND_FETCH_DIM_RW
 				|| opline->opcode == ZEND_FETCH_DIM_FUNC_ARG
 				|| opline->opcode == ZEND_FETCH_DIM_UNSET)) {
 			opline->extended_value = ZEND_FETCH_DIM_OBJ;
 		}
 
+		/* ASSIGN_OBJ and FETCH_OBJ_[R]W are missing handlers for CONST|TMP op1. */
+		bool obj_node_was_tmp_var = obj_node.op_type == IS_TMP_VAR;
+		if ((type == BP_VAR_W || type == BP_VAR_RW)) {
+			if (obj_node.op_type == IS_TMP_VAR) {
+				obj_node.op_type = IS_VAR;
```

Comment:
How it's possible that we perform "WRITE" mode fetching into TMP_VAR in the first place?
This shouldn't work, as this is at least missing dereferencing.
If this patch intends to change only the FETCH_CONST instruction, it's better to do this in a more obvious way.
May be I didn't understand the patch yet.

Reply:
> How it's possible that we perform "WRITE" mode fetching into TMP_VAR in the first place?

@dstogov It wasn't possible so far. With this PR, `FOO->bar = 'bar';` leads to this scenario (`T1 = FETCH_CONST "FOO"; T2 = ASSIGN_OBJ T1 "BAR"; DATA "bar"` ) as `FETCH_CONST` returns a `TMP`.

>  This shouldn't work, as this is at least missing dereferencing.

Can you clarify? The lhs of `->` is neither a reference nor an indirect value in this case, as it is the value of the constant, i.e. a `IS_OBJ` zval.

>  If this patch intends to change only the FETCH_CONST instruction, it's better to do this in a more obvious way.

I can think of three options:

* Change the result of `FETCH_CONST` to `VAR` only where necessary (what I do in this PR).
* Add `TMP` to OP1 in all the relevant handlers (`ZEND_ASSIGN_OBJ`, `ZEND_FETCH_OBJ_W`, etc). With `TMPVAR` this shouldn't actually change the size of the VM, or impact performance.
* Change the result of all `FETCH_CONST` to `VAR`. This would have a negative performance impact due to added reference/indirect checks when using the result.

I might be missing something obvious, please let me know what your preferred approach is.